src/App.tsx

```diff
@@ -1,26 +1,304 @@
+  const [isAllTodosActive, setIsAllTodosActive] = useState(false);
+  const [isCheckLoading, setIsCkeckLoading] = useState(-1);
+  const [errorTimeout, setErrorTimeout] = useState<NodeJS.Timeout | null>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    inputRef.current?.focus();
+  }, [isInputDisabled, loadingTodos]);
+
+  useEffect(() => {
+    setIsAllTodosActive(todos.every(todo => todo.completed));
+  }, [todos]);
+
+  const updateClearCompletedState = useCallback((todosList: Todo[]) => {
+    setIsClearCompletedDesabled(!todosList.some(todo => todo.completed));
+  }, []);
+
+  const showError = useCallback(
+    (message: string) => {
+      setErrors(prev => ({
+        messages: [...new Set([...prev.messages, message])],
+      }));
+      setNotificationVisible(true);
+
+      if (errorTimeout) {
+        clearTimeout(errorTimeout);
+      }
+
+      const timeout = setTimeout(() => {
+        setErrors({ messages: [] });
+        setNotificationVisible(false);
+      }, 3000);
+
+      setErrorTimeout(timeout);
+    },
+    [errorTimeout],
+  );
+
+  const handleChangeCompleted = async (todo: Todo) => {
+    setIsCkeckLoading(todo.id);
+    try {
+      const editedTodo = { ...todo, completed: !todo.completed };
+
+      await editTodo(editedTodo);
+      setTodos(prevTodos =>
+        prevTodos.map(t => (t.id === editedTodo.id ? editedTodo : t)),
+      );
+    } catch (error) {
+      showError('Unable to update a todo');
+    } finally {
+      setIsCkeckLoading(-1);
+    }
+  };
+
+  const handleFormSubmit = async (event: React.FormEvent) => {
+    event.preventDefault();
+
+    if (input.trim().length === 0) {
+      showError('Title should not be empty');
+
+      return;
+    }
+
+    setIsInputDisabled(true);
+
+    const temporaryTodo: Todo = {
+      id: 0,
+      userId: USER_ID,
+      title: input.trim(),
+      completed: false,
+    };
+
+    setTempTodo(temporaryTodo);
+
+    try {
+      const newTodo = await addTodo({
+        userId: USER_ID,
+        title: input.trim(),
+        completed: false,
+      });
+
+      setTodos(prevTodos => {
+        const updatedTodos = [...prevTodos, newTodo];
+
+        updateClearCompletedState(updatedTodos);
+
+        return updatedTodos;
+      });
+      setTempTodo(null);
+      setInput('');
+    } catch (error) {
+      showError('Unable to add a todo');
+      setTempTodo(null);
+    } finally {
+      setIsInputDisabled(false);
+      inputRef.current?.focus();
+    }
+  };
+
+  const handleDeleteCompletedTodos = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    if (completedTodos.length === 0) {
+      return;
+    }
+
+    const failedDeletions: Todo[] = [];
+
+    try {
+      await Promise.all(
+        completedTodos.map(async todo => {
+          try {
+            await deleteTodo(todo.id);
+          } catch {
+            failedDeletions.push(todo);
```

Comment:
If you need an array of failed requests use `Promise.allsettled` instead https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled